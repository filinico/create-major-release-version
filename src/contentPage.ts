import fs from 'fs'
import path from 'path'

export const adaptContentForMajorVersion = (
  workspace: string,
  contentPath: string,
  releaseVersion: string
): string => {
  const filePath = path.resolve(workspace, contentPath)
  const rawData = fs.readFileSync(filePath, 'utf8')
  return rawData.replace(/{{RELEASE_VERSION}}/g, releaseVersion)
}
