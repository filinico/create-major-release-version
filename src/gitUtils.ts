import * as core from '@actions/core'
import {exec} from 'promisify-child-process'

export const gotoDirectory = async (directoryPath: string): Promise<void> => {
  const {stderr} = await exec(`cd ${directoryPath}`)
  if (stderr) {
    core.error(stderr.toString())
  }
}

export const createDirectory = async (directoryName: string): Promise<void> => {
  const {stderr} = await exec(`mkdir ${directoryName}`)
  if (stderr) {
    core.error(stderr.toString())
  }
}

export const createBranch = async (releaseVersion: string): Promise<void> => {
  const {stderr} = await exec(
    `git checkout -b release/${releaseVersion} develop`
  )
  if (stderr) {
    core.error(stderr.toString())
  }
}

export const doesBranchExist = async (
  releaseVersion: string
): Promise<boolean> => {
  const {stderr, stdout} = await exec(
    `git ls-remote origin release/${releaseVersion}`
  )
  if (stderr) {
    core.error(stderr.toString())
  }
  if (stdout) {
    return true
  } else {
    return false
  }
}

export const commit = async (commitMessage: string): Promise<void> => {
  await exec(`git add .`)
  const {stderr} = await exec(`git commit -m "${commitMessage}"`)
  if (stderr) {
    core.error(stderr.toString())
  }
}

export const push = async (): Promise<void> => {
  const {stderr} = await exec(`git push`)
  if (stderr) {
    core.error(stderr.toString())
  }
}
