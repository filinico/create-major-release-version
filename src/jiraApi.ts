import * as core from '@actions/core'
import axios from 'axios'

export interface JiraContext {
  subDomain: string
  email: string
  token: string
  projectsIds: string[]
  projectsKeys: string[]
  masterProjectId: string
  masterProjectKey: string
  masterIssueType: string
}

export interface JiraVersion {
  self?: string
  id?: string
  name: string
  archived: boolean
  released: boolean
  startDate?: string
  releaseDate?: string
  userStartDate?: string
  userReleaseDate?: string
  projectId?: number
  description?: string
  overdue?: boolean
}

interface AuthHeaders {
  headers: {
    Authorization: string
    Accept: string
  }
}

export interface SearchedJiraIssue {
  expand?: string
  id?: string
  self?: string
  key: string
  fields: {
    summary?: string
  }
}

export interface CreatedIssue {
  id: string
  key: string
  self: string
  transition: {
    status: number
    errorCollection: {
      errorMessages: string[]
      errors: {}
      status: number
    }
  }
}

export interface JiraIssue {
  update: {}
  fields?: JiraFields
}

interface JiraIssueDescriptionContent {
  type: string
  attrs?: {
    isNumberColumnEnabled?: boolean
    layout?: string
    url?: string
  }
  content?: (JiraIssueDescriptionContent | JiraIssueDescriptionText)[]
}

interface JiraIssueDescriptionText {
  type: string
  text: string
  marks?: {
    type: string
    attrs?: {
      href: string
    }
  }[]
}

interface JiraCustomField {
  id?: string
  value?: string
  child?: {
    value: string
  }
}

interface JiraFields {
  summary?: string
  issuetype?: {
    id: string
  }
  project?: {
    id: string
  }
  description?: {
    type: string
    version: number
    content: JiraIssueDescriptionContent[]
  }
  fixVersions?: {
    id: string
  }[]
  customfield_23944?: JiraCustomField
  customfield_23710?: JiraCustomField
  customfield_21603?: JiraCustomField
  customfield_12803?: JiraCustomField
}

export const listProjectVersions = async (
  context: JiraContext,
  projectKey: string
): Promise<JiraVersion[]> => {
  const {subDomain, email, token} = context
  try {
    core.info(`request listProjectVersions ${projectKey}`)
    const response = await axios.get(
      `https://${subDomain}.atlassian.net/rest/api/3/project/${projectKey}/versions`,
      getAuthHeaders(email, token)
    )
    core.info(`listProjectVersions successful`)
    return response?.data
  } catch (error: unknown) {
    core.error('error during listProjectVersions request')
    if (axios.isAxiosError(error)) {
      core.error(error.message)
      core.error(JSON.stringify(error.toJSON))
    }
    return Promise.reject(error)
  }
}

export const createVersion = async (
  context: JiraContext,
  version: JiraVersion
): Promise<JiraVersion> => {
  const {subDomain, email, token} = context
  try {
    core.info('request createVersion')
    const response = await axios.post(
      `https://${subDomain}.atlassian.net/rest/api/3/version`,
      version,
      getAuthHeaders(email, token)
    )
    core.info(`createVersion successful`)
    return response?.data
  } catch (error: unknown) {
    core.error('error during createVersion request')
    if (axios.isAxiosError(error)) {
      core.error(error.message)
      core.error(JSON.stringify(error.toJSON))
    }
    return Promise.reject(error)
  }
}

export const searchIssues = async (
  context: JiraContext,
  jQLQuery: string,
  properties: string[]
): Promise<SearchedJiraIssue[]> => {
  const {subDomain, email, token} = context
  try {
    core.info('request searchIssues')
    const response = await axios.post(
      `https://${subDomain}.atlassian.net/rest/api/3/search`,
      {
        jql: jQLQuery,
        maxResults: 100,
        fieldsByKeys: true,
        fields: properties,
        startAt: 0
      },
      getAuthHeaders(email, token)
    )
    core.info(`searchIssues successful`)
    let issues: SearchedJiraIssue[] = []
    if (response?.data?.issues && response?.data?.issues.length > 0) {
      issues = response.data.issues
    }
    return issues
  } catch (error: unknown) {
    core.error('error during searchIssues request')
    if (axios.isAxiosError(error)) {
      core.error(error.message)
      core.error(JSON.stringify(error.toJSON))
    }
    return Promise.reject(error)
  }
}

export const createIssue = async (
  context: JiraContext,
  data: JiraIssue
): Promise<CreatedIssue> => {
  const {subDomain, email, token} = context
  try {
    core.info('request createIssue')
    core.info(`createIssue:${JSON.stringify(data)}`)
    const response = await axios.post(
      `https://${subDomain}.atlassian.net/rest/api/3/issue`,
      data,
      getAuthHeaders(email, token)
    )
    core.info(`createIssue successful`)
    return response?.data
  } catch (error: unknown) {
    core.error('error during createIssue request')
    if (axios.isAxiosError(error)) {
      core.error(error.message)
      core.error(JSON.stringify(error.toJSON))
    }
    return Promise.reject(error)
  }
}

const getAuthHeaders = (email: string, token: string): AuthHeaders => {
  return {
    headers: {
      Authorization: `Basic ${Buffer.from(`${email}:${token}`).toString(
        'base64'
      )}`,
      Accept: 'application/json'
    }
  }
}
