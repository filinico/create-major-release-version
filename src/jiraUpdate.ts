import * as core from '@actions/core'
import {
  CreateConfluencePage,
  JiraContext,
  JiraVersion,
  createConfluencePage,
  createIssue,
  createVersion,
  listProjectVersions,
  searchIssues
} from './jiraApi'

export const configureJira = async (
  jiraContext: JiraContext,
  releaseVersion: string,
  tagPrefix: string
): Promise<void> => {
  const majorVersion = `${tagPrefix}${releaseVersion}.0`
  await createVersionsOfProjects(jiraContext, majorVersion)
  const {masterProjectId, masterProjectKey, masterIssueType} = jiraContext
  const masterTicketVersion = await createIfNotExistsJiraVersion(
    jiraContext,
    majorVersion,
    parseInt(masterProjectId),
    masterProjectKey
  )

  if (masterTicketVersion && masterTicketVersion.id) {
    core.info(
      `request creation of master ticket version ${majorVersion} with id  ${masterTicketVersion.id}`
    )
    await createMasterTicket(
      majorVersion,
      masterIssueType,
      masterProjectId,
      masterTicketVersion.id,
      jiraContext
    )
  }
}

const createVersionsOfProjects = async (
  jiraContext: JiraContext,
  majorVersion: string
): Promise<void> => {
  const {projectsIds, projectsKeys} = jiraContext
  for (let i = 0; i < projectsKeys.length; i++) {
    const projectId = projectsIds[i]
    const projectKey = projectsKeys[i]
    await createIfNotExistsJiraVersion(
      jiraContext,
      majorVersion,
      parseInt(projectId),
      projectKey
    )
  }
}

export const createIfNotExistsJiraVersion = async (
  context: JiraContext,
  fixVersion: string,
  projectId: number,
  projectKey: string
): Promise<JiraVersion> => {
  const versions = await listProjectVersions(context, projectKey)
  const result = versions.filter(i => i.name === fixVersion)
  let version: JiraVersion
  if (!result || result.length === 0) {
    const requestedVersion: JiraVersion = {
      name: fixVersion,
      archived: false,
      released: false,
      projectId
    }
    core.info(`version not found. start create version:[${requestedVersion}]`)
    version = await createVersion(context, requestedVersion)
    core.info(`version created:[${version.id}]`)
  } else {
    version = result[0]
    core.info(`version found:[${version.id}]`)
  }
  return version
}

export const createMasterTicket = async (
  version: string,
  masterIssueType: string,
  masterProjectId: string,
  masterTicketVersionId: string,
  jiraContext: JiraContext
): Promise<void> => {
  const masterTicket = await getMasterTicketKey(jiraContext, version)
  if (!masterTicket) {
    await createIssue(jiraContext, {
      update: {},
      fields: {
        summary: `${version} Master Ticket`,
        issuetype: {
          id: masterIssueType
        },
        project: {
          id: masterProjectId
        },
        description: {
          type: 'doc',
          version: 1,
          content: [
            {
              type: 'paragraph',
              content: [
                {
                  text: 'Not released yet.',
                  type: 'text'
                }
              ]
            }
          ]
        },
        fixVersions: [
          {
            id: masterTicketVersionId
          }
        ],
        customfield_23944: {
          value: 'Major'
        },
        customfield_23710: {
          value: 'Power App',
          child: {
            value: 'Treasury Management (CTM)'
          }
        },
        customfield_21603: {
          value: 'Treasury Management (CTM)'
        },
        customfield_12803: {
          id: masterTicketVersionId
        }
      }
    })
  }
}

export const getMasterTicketKey = async (
  context: JiraContext,
  fixVersion: string
): Promise<string | null> => {
  const {masterProjectKey} = context
  const masterTicketQuery = `project = ${masterProjectKey} AND fixVersion in (${fixVersion})`
  core.info(`masterTicketQuery:[${masterTicketQuery}]`)
  const issues = await searchIssues(context, masterTicketQuery, ['summary'])
  let masterTicketIssueKey: string | null = null
  if (issues && issues.length === 1) {
    masterTicketIssueKey = issues[0].key
  }
  core.info(`masterTicketIssueKey:${masterTicketIssueKey}`)
  return masterTicketIssueKey
}

export const createReleaseContentPage = async (
  context: JiraContext,
  releaseVersion: string,
  contentPage: string
): Promise<void> => {
  const {subDomain, confluenceSpaceKey, ancestorPage} = context
  const confluencePage: CreateConfluencePage = {
    title: `Release content of major version ${releaseVersion}`,
    type: 'page',
    space: {
      key: confluenceSpaceKey
    },
    ancestors: [
      {
        id: parseInt(ancestorPage)
      }
    ],
    body: {
      storage: {
        representation: 'storage',
        value: contentPage
      }
    }
  }

  const createdPage = await createConfluencePage(context, confluencePage)
  core.info(
    `created confluence page: https://${subDomain}.atlassian.net/wiki${createdPage._links.webui}`
  )
}
