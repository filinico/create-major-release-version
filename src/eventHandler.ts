import * as core from '@actions/core'
import * as github from '@actions/github'
import {
  addAuthor,
  commit,
  createBranch,
  diff,
  doesBranchExist,
  fetch,
  gotoDirectory,
  mergeIntoCurrent,
  push
} from './gitUtils'
import {
  applyNextVersion,
  getNextVersion,
  getPreviousVersion,
  getVersionFromTag
} from './version'
import {
  configureSettings,
  getVersionsFromSettings,
  loadCodeOwners,
  writeCodeOwners
} from './settings'
import {configureWorkflow, configureWorkflowPreviousRelease} from './workflows'
import {mergePullRequest, openPullRequest} from './gitHubApi'
import {Context} from '@actions/github/lib/context'
import {JiraContext} from './jiraApi'
import {configureJira} from './jiraUpdate'
import {configureScripts} from './scripts'
type GitHub = ReturnType<typeof github.getOctokit>

export interface GitHubContext {
  octokit: GitHub
  context: Context
  workspace: string
  settingsPath: string
  versionPrefix: string
  tagPrefix: string
  gitEmail: string
  gitUser: string
  workflowPath: string
  versionPath: string
  scriptsPath: string
}

interface ReleaseInfo {
  currentRelease: string
  nextRelease: string
}

export const onReleaseCreated = async (
  actionContext: GitHubContext,
  jiraContext: JiraContext
): Promise<ReleaseInfo> => {
  const {context, workspace, tagPrefix, gitEmail, gitUser, settingsPath} =
    actionContext
  const {
    payload: {
      release: {tag_name, target_commitish, prerelease, id}
    },
    sha
  } = context
  core.info(`tag_name:${tag_name}`)
  core.info(`target_commitish:${target_commitish}`)
  core.info(`prerelease:${prerelease}`)
  core.info(`id:${id}`)
  core.info(`revision:${sha}`)
  const releaseVersion = getVersionFromTag(tagPrefix, tag_name)
  core.info(`Release version:${releaseVersion}`)
  const releaseBranch = `release/${releaseVersion}`
  core.info(`Release branch:${releaseBranch}`)
  const previousVersion = getPreviousVersion(releaseVersion)
  core.info(`Previous version:${previousVersion}`)
  const previousReleaseBranch = `release/${previousVersion}`
  core.info(`Previous release branch:${previousReleaseBranch}`)
  if (!tag_name.includes('.0.0')) {
    throw new Error(
      `Release branch ${releaseBranch} is not a major version ending with .0.0`
    )
  }
  await gotoDirectory(workspace)
  await fetch(target_commitish)
  await fetch(previousReleaseBranch)
  const releaseBranchExists = await doesBranchExist(releaseBranch)
  const conflictsExists = await diff(
    previousReleaseBranch,
    target_commitish,
    settingsPath
  )
  if (releaseBranchExists || conflictsExists) {
    if (releaseBranchExists) {
      core.error(`Release branch ${releaseBranch} already exists`)
    }
    if (conflictsExists) {
      core.error(
        `There are conflicts between the release branch ${releaseBranch} and ${target_commitish}. Please resolve the conflicts and create a new GitHub release.`
      )
    }
    throw new Error(
      `Cannot proceed with the creation of the release branch ${releaseBranch}.`
    )
  }

  await addAuthor(gitEmail, gitUser)
  core.info(`Author identity added`)
  await configurePreviousVersion(
    actionContext,
    releaseVersion,
    previousVersion,
    previousReleaseBranch
  )
  await createNewMajorVersion(
    actionContext,
    releaseVersion,
    releaseBranch,
    previousVersion,
    previousReleaseBranch
  )
  await configureNextVersion(
    actionContext,
    releaseVersion,
    previousVersion,
    releaseBranch
  )
  await configureJira(jiraContext, releaseVersion, tagPrefix)
  const currentRelease = releaseVersion.replace('.0', '')
  const nextRelease = getNextVersion(releaseVersion).replace('.0', '')
  return {
    currentRelease,
    nextRelease
  }
}

const createNewMajorVersion = async (
  actionContext: GitHubContext,
  releaseVersion: string,
  releaseBranch: string,
  previousVersion: string,
  previousReleaseBranch: string
): Promise<void> => {
  const {context, workspace, settingsPath, versionPrefix, workflowPath} =
    actionContext
  const {
    payload: {
      release: {target_commitish}
    }
  } = context
  core.info(`Start creation of new major version`)
  await fetch(previousReleaseBranch)
  await fetch(target_commitish)
  core.info(`fetch successful`)
  const codeOwners = loadCodeOwners(workspace)
  await createBranch(releaseBranch, target_commitish)
  core.info(`Release branch created`)
  await mergeIntoCurrent(previousReleaseBranch, releaseBranch)
  core.info(`Previous release branch merged`)
  writeCodeOwners(workspace, codeOwners)
  configureSettings(
    releaseVersion,
    workspace,
    settingsPath,
    versionPrefix,
    target_commitish
  )
  configureWorkflow(releaseVersion, workspace, workflowPath, target_commitish)
  await commit(`setup new version ${releaseVersion}`)
  core.info(`changes committed`)
  await push()
  core.info(`changes pushed`)
  core.info(`New major version created`)
}

const configurePreviousVersion = async (
  actionContext: GitHubContext,
  releaseVersion: string,
  previousVersion: string,
  previousReleaseBranch: string
): Promise<void> => {
  const {workspace, workflowPath} = actionContext
  core.info(`Start configuration of previous version`)
  await fetch(previousReleaseBranch)
  core.info(`fetch successful`)
  const configurationBranch = `automation/configure-previous-version-${previousVersion}`
  await createBranch(configurationBranch, previousReleaseBranch)
  configureWorkflowPreviousRelease(releaseVersion, workspace, workflowPath)
  await commit(`configure new version ${releaseVersion} on ${previousVersion}`)
  core.info(`changes committed`)
  await push()
  core.info(`changes pushed`)
  const title = `Configure new version ${releaseVersion} on ${previousVersion}`
  const pullRequestId = await openPullRequest(
    actionContext,
    title,
    title,
    configurationBranch,
    previousReleaseBranch
  )
  const mergeCommitMessage = `Merge pull request #${pullRequestId} from ${configurationBranch}`
  const isMerged = await mergePullRequest(
    actionContext,
    pullRequestId,
    mergeCommitMessage,
    mergeCommitMessage
  )
  if (isMerged) {
    core.info(`PR merged on previous version`)
  } else {
    core.error(`PR not merged on previous version`)
  }
  core.info(`Previous version configured`)
}

const configureNextVersion = async (
  actionContext: GitHubContext,
  releaseVersion: string,
  previousVersion: string,
  releaseBranch: string
): Promise<void> => {
  const {context, workspace, settingsPath, versionPath, scriptsPath} =
    actionContext
  const {
    payload: {
      release: {target_commitish}
    }
  } = context
  core.info(`Start configuration of next version`)
  await fetch(releaseBranch)
  await fetch(target_commitish)
  core.info(`fetch successful`)
  const nextVersion = getNextVersion(releaseVersion)
  const configurationBranch = `automation/configure-next-version-${nextVersion}`
  await createBranch(configurationBranch, target_commitish)
  const codeOwners = loadCodeOwners(workspace)
  await mergeIntoCurrent(releaseBranch, configurationBranch)
  core.info(`Release branch merged into ${target_commitish}`)
  writeCodeOwners(workspace, codeOwners)
  const {nextDbVersion, currentDbVersion, nextArtifactVersion} =
    getVersionsFromSettings(workspace, settingsPath, target_commitish)
  applyNextVersion(nextArtifactVersion, workspace, versionPath)
  core.info(`Next version modified to ${nextDbVersion}`)
  await configureScripts(
    currentDbVersion,
    nextDbVersion,
    workspace,
    scriptsPath
  )
  core.info(`Scripts added for next version ${nextDbVersion}`)
  await commit(`configure next version ${nextVersion} on ${target_commitish}`)
  core.info(`Next version changes committed`)
  await push()
  core.info(`Next version changes pushed`)
  const title = `Configure next version ${nextVersion} on ${target_commitish}`
  const pullRequestId = await openPullRequest(
    actionContext,
    title,
    title,
    configurationBranch,
    target_commitish
  )
  const mergeCommitMessage = `Merge pull request #${pullRequestId} from ${configurationBranch}`
  const isMerged = await mergePullRequest(
    actionContext,
    pullRequestId,
    mergeCommitMessage,
    mergeCommitMessage
  )
  if (isMerged) {
    core.info(`PR merged on ${target_commitish}`)
  } else {
    core.error(`PR not merged on ${target_commitish}`)
  }
  core.info(`Next version configured on ${target_commitish}`)
}
