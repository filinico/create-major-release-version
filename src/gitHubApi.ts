import * as core from '@actions/core'
import {GitHubContext} from './eventHandler'

export const openPullRequest = async (
  actionContext: GitHubContext,
  title: string,
  body: string,
  sourceBranch: string,
  targetBranch: string
): Promise<number> => {
  const {octokit, context} = actionContext
  const {
    repo: {repo, owner}
  } = context
  try {
    const {data: pullRequest} = await octokit.rest.pulls.create({
      owner,
      repo,
      title,
      body,
      head: sourceBranch,
      base: targetBranch,
      draft: false
    })
    return pullRequest.number
  } catch (err: unknown) {
    core.error(
      `There was an error during the opening of pull request ${sourceBranch} to ${targetBranch}`
    )
    const error = err as string
    core.error(error)
    return 0
  }
}

export const mergePullRequest = async (
  actionContext: GitHubContext,
  pullRequestId: number,
  mergeCommitMessage: string,
  mergeCommitDescription: string
): Promise<boolean> => {
  const {octokit, context} = actionContext
  const {
    repo: {repo, owner}
  } = context
  try {
    const {
      data: {merged}
    } = await octokit.rest.pulls.merge({
      owner,
      repo,
      pull_number: pullRequestId,
      commit_message: mergeCommitDescription,
      commit_title: mergeCommitMessage,
      merge_method: 'merge'
    })
    return merged
  } catch (err: unknown) {
    core.error(
      `There was an error during the merge of pull request ${pullRequestId}`
    )
    const error = err as string
    core.error(error)
    return false
  }
}

export const createProject = async (
  actionContext: GitHubContext,
  name: string,
  description: string
): Promise<number> => {
  const {octokit, context} = actionContext
  const {
    repo: {repo, owner}
  } = context
  const {
    data: {id}
  } = await octokit.rest.projects.createForRepo({
    owner,
    repo,
    name,
    body: description
  })
  return id
}

export const createProjectColumn = async (
  actionContext: GitHubContext,
  projectId: number,
  name: string
): Promise<number> => {
  const {octokit, context} = actionContext
  const {
    repo: {repo, owner}
  } = context
  const {
    data: {id}
  } = await octokit.rest.projects.createColumn({
    owner,
    repo,
    project_id: projectId,
    name
  })
  return id
}
