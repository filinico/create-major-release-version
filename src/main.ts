import * as core from '@actions/core'
import * as github from '@actions/github'
import {onReleaseCreated} from './eventHandler'

async function run(): Promise<void> {
  try {
    const githubToken = core.getInput('GITHUB_TOKEN', {required: true})
    const settingsPath = core.getInput('SETTINGS_FILE', {required: true})
    const versionPrefix = core.getInput('VERSION_PREFIX', {required: true})
    const tagPrefix = core.getInput('TAG_PREFIX', {required: true})
    const gitEmail = core.getInput('GIT_USER_EMAIL', {required: true})
    const gitUser = core.getInput('GIT_USER_NAME', {required: true})
    const workflowPath = core.getInput('WORKFLOW_FILE', {required: true})
    const versionPath = core.getInput('VERSION_FILE', {required: true})
    const scriptsPath = core.getInput('SCRIPTS_PATH', {required: true})

    core.info(`GITHUB workspace=${process.env.GITHUB_WORKSPACE}`)

    if (process.env.GITHUB_WORKSPACE === undefined) {
      throw new Error('GITHUB_WORKSPACE not defined.')
    }

    const octokit = github.getOctokit(githubToken)
    const gitHubContext = {
      octokit,
      context: github.context,
      workspace: process.env.GITHUB_WORKSPACE,
      settingsPath,
      versionPrefix,
      tagPrefix,
      gitUser,
      gitEmail,
      workflowPath,
      versionPath,
      scriptsPath
    }

    const jiraContext = {
      subDomain: core.getInput('JIRA_SUBDOMAIN', {required: true}),
      email: core.getInput('JIRA_USER', {required: true}),
      token: core.getInput('JIRA_TOKEN', {required: true}),
      projectsIds: core
        .getInput('JIRA_PROJECTS_IDS', {required: true})
        .split(','),
      projectsKeys: core
        .getInput('JIRA_PROJECTS_KEYS', {required: true})
        .split(','),
      masterProjectId: core.getInput('JIRA_MASTER_PROJECT_ID', {
        required: true
      }),
      masterProjectKey: core.getInput('JIRA_MASTER_PROJECT_KEY', {
        required: true
      })
    }

    core.info(`GITHUB_EVENT_NAME=${process.env.GITHUB_EVENT_NAME}`)
    core.info(`GITHUB context action=${gitHubContext.context.payload.action}`)
    if (
      process.env.GITHUB_EVENT_NAME === 'release' &&
      github.context.payload.action === 'created'
    ) {
      core.info(`start onReleaseCreated`)
      const releaseInfo = await onReleaseCreated(gitHubContext, jiraContext)
      core.setOutput('RELEASE_INFO', releaseInfo)
      core.info(`onReleaseCreated finished`)
    }
  } catch (error) {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    core.setFailed(error.message)
  }
}

run()
