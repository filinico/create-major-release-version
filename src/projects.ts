import {createProject, createProjectColumn} from './gitHubApi'
import {GitHubContext} from './eventHandler'
import {configureAssignProjectWorkflow} from './workflows'

export interface ProjectBoard {
  inProgressColumnId: number
  doneColumnId: number
}

export const configureProjects = async (
  actionContext: GitHubContext,
  releaseVersion: string
): Promise<void> => {
  const {inProgressColumnId} = await createProjectBoard(
    actionContext,
    releaseVersion
  )
  const {workspace, assignProjectPath} = actionContext
  configureAssignProjectWorkflow(
    workspace,
    assignProjectPath,
    releaseVersion,
    inProgressColumnId.toString()
  )
  // TODO: configure workflow archive project cards
}

const createProjectBoard = async (
  actionContext: GitHubContext,
  releaseVersion: string
): Promise<ProjectBoard> => {
  const projectName = `${releaseVersion} Release Tracker`
  const projectDescription = `Progress overview of pull requests targeting the release branch ${releaseVersion}`
  const projectId = await createProject(
    actionContext,
    projectName,
    projectDescription
  )
  const inProgressColumnId = await createProjectColumn(
    actionContext,
    projectId,
    'In progress'
  )
  await createProjectColumn(actionContext, projectId, 'Review in progress')
  await createProjectColumn(actionContext, projectId, 'Reviewer approved')
  const doneColumnId = await createProjectColumn(
    actionContext,
    projectId,
    'Done'
  )
  return {
    inProgressColumnId,
    doneColumnId
  }
}
