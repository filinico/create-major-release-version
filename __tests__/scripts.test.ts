import {
  applyVersionsIntoFile,
  configureScripts,
  listFiles
} from '../src/scripts'
import path from 'path'
import fs from 'fs'
import {removeDirectory} from '../src/gitUtils'

jest.setTimeout(70000)
const workspace = './__tests__/testData'
const directoryPath = 'scripts'
const scriptsPath = path.resolve(workspace, directoryPath)

beforeEach(async () => {})

test('list files of directory recursive', async () => {
  const templatesPath = path.resolve(scriptsPath, 'templates')
  expect(listFiles(templatesPath)).toEqual([
    path.resolve(templatesPath, 'secondLevel', 'XX_anotherScript.sql'),
    path.resolve(templatesPath, 'XX_scriptTemplate.sql')
  ])
})

test('apply versions into file', async () => {
  const templatePath = path.resolve(
    scriptsPath,
    'templates',
    'XX_scriptTemplate.sql'
  )
  const rawData = fs.readFileSync(templatePath, 'utf8')
  const filePath = path.resolve(scriptsPath, 'script.sql')
  fs.writeFileSync(filePath, rawData)
  applyVersionsIntoFile(filePath, 'v.0.010', 'v.0.011')
  const newData = fs.readFileSync(filePath, 'utf8')
  expect(newData).toMatchSnapshot()
})

test('configure scripts', async () => {
  await configureScripts('v.0.010', 'v.0.011', workspace, directoryPath)
  const files = listFiles(path.resolve(scriptsPath, 'v.0.011'))
  expect(files).toHaveLength(2)
  for (const file of files) {
    const newData = fs.readFileSync(file, 'utf8')
    expect(newData).toMatchSnapshot()
    const filename = file
      .substring(file.lastIndexOf('\\') + 1)
      .substring(file.lastIndexOf('/') + 1)
    expect(filename.startsWith('XX')).toEqual(false)
    expect(filename.startsWith('v011')).toEqual(true)
  }
  await removeDirectory(path.resolve(scriptsPath, 'v.0.011'))
})
