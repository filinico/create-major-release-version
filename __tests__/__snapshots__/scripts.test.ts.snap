// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apply versions into file 1`] = `
"DECLARE @DBVersion AS nvarchar(10), @ExpectedDBVersion AS nvarchar(10), @NextDBVersion AS nvarchar(10), @NextAppVersion as nvarchar(10);
SET @ExpectedDBVersion = 'v.0.010';
SET @NextDBVersion = 'v.0.011';
SET @NextAppVersion = 'v.110';
PRINT '+--------------------------------------------------------------+';
PRINT '| current database version: ' + @DBVersion + '                            |';
PRINT '+--------------------------------------------------------------+';
PRINT '';
PRINT '########  ########'
PRINT '';
SET @NextAppVersion = 'v.110';
SET @ExpectedDBVersion = 'v.0.010';
SET @NextDBVersion = 'v.0.011';
PRINT '';
PRINT '########  ########'
PRINT '';"
`;

exports[`configure scripts 1`] = `
"DECLARE @DBVersion AS nvarchar(10), @ExpectedDBVersion AS nvarchar(10), @NextDBVersion AS nvarchar(10);
SET @ExpectedDBVersion = 'v.0.010';
SET @NextDBVersion = 'v.0.011';
PRINT '+--------------------------------------------------------------+';
PRINT '| current database version: ' + @DBVersion + '                            |';
PRINT '+--------------------------------------------------------------+';
PRINT '';
PRINT '########  ########'
PRINT '';
SET @ExpectedDBVersion = 'v.0.010';
SET @NextDBVersion = 'v.0.011';
PRINT '';
PRINT '########  ########'
PRINT '';"
`;

exports[`configure scripts 2`] = `
"DECLARE @DBVersion AS nvarchar(10), @ExpectedDBVersion AS nvarchar(10), @NextDBVersion AS nvarchar(10), @NextAppVersion as nvarchar(10);
SET @ExpectedDBVersion = 'v.0.010';
SET @NextDBVersion = 'v.0.011';
SET @NextAppVersion = 'v.110';
PRINT '+--------------------------------------------------------------+';
PRINT '| current database version: ' + @DBVersion + '                            |';
PRINT '+--------------------------------------------------------------+';
PRINT '';
PRINT '########  ########'
PRINT '';
SET @NextAppVersion = 'v.110';
SET @ExpectedDBVersion = 'v.0.010';
SET @NextDBVersion = 'v.0.011';
PRINT '';
PRINT '########  ########'
PRINT '';"
`;
